toponetx/classes/path_complex.py

```diff
@@ -0,0 +1,1149 @@
+                        sub_path = tuple(sub_path)
+
+                        # add to _G
+                        if len(sub_path) == 1:
+                            self._G.add_node(sub_path[0])
+                        elif len(sub_path) == 2:
+                            self._G.add_edge(sub_path[0], sub_path[1])
+
+                        # expand _path_set if necessary. keep track of newly added paths to expend _allowed_paths
+                        new_path = self._update_faces_dict_entry(sub_path)
+                        if new_path is not None:
+                            new_paths.add(new_path)
+                # update allowed paths
+                if len(new_paths) > 0:
+                    self._allowed_paths.update(new_paths)
+
+                self._update_attributes(path, **attr)
+
+    @property
+    def dim(self) -> int:
+        """Dimension.
+
+        Returns
+        -------
+        int
+            This is the highest dimension of any elementary p-path in the complex.
+        """
+        return self._path_set.max_dim
+
+    @property
+    def nodes(self) -> NodeView:
+        """Nodes.
+
+        Returns
+        -------
+        NodeView
+            A view of all nodes in the path complex.
+        """
+        return self._G.nodes
+
+    @property
+    def edges(self) -> EdgeView:
+        """Edges.
+
+        Returns
+        -------
+        EdgeView
+            A view of all edges in the path complex.
+        """
+        return self._G.edges
+
+    @property
+    def paths(self) -> PathView:
+        """
+        Set of all elementary p-paths.
+
+        Returns
+        -------
+        PathView
+            A view of all elementary p-paths in the path complex.
+        """
+        return self._path_set
+
+    @property
+    def shape(self) -> tuple[int, ...]:
+        """Shape of path complex.
+
+        (number of elementary p-paths for each dimension d, for d in range(0,dim(Pc)))
+
+        Returns
+        -------
+        tuple of ints
+        """
+        return self._path_set.shape
+
+    def clone(self) -> "PathComplex":
+        """Return a copy of the path complex.
+
+        The clone method by default returns an independent shallow copy of the path complex. Use Python’s
+        `copy.deepcopy` for new containers.
+
+        Returns
+        -------
+        PathComplex
+        """
+        return PathComplex(list(self.paths), name=self.name)
+
+    def skeleton(self, rank: int) -> set[tuple[Hashable]]:
+        """Compute skeleton.
+
+        Returns
+        -------
+        set[tuple[Hashable]]
+            Set of elementary p-paths of dimension specified by `rank`.
+        """
+        if rank < len(self._path_set.faces_dict) and rank >= 0:
+            tmp = (path for path in self._path_set.faces_dict[rank].keys())
+            return sorted(
+                tmp, key=lambda x: tuple(map(str, x))
+            )  # lexicographic comparison
```

Comment:
This doesn't conform with the function signature and docstring.

Reply:
@quang-truong I am not sure if you have addressed this one, if so please mark it as resolved. Thanks!

Reply:
@mhajij @ffl096 i just updated the code for `skeleton()` and added extra test cases. Please refer to the latest commit for changes.